### DISCUSSION TITLE
parallel execution support for behave

### DISCUSSION THREAD

**Discussion by joshij1979** [behave/behave#930] [General]
Is there any plan to support parallel execution natively in behave, just like pytest has it?

My use case is to use parallel testing for features on a build server / locally to reduce the test execution time.  Also would like to hear how other in the community are approaching this type of use case.

**Comment by xrg** [comment]
https://github.com/xrg/behave-parallel

**Comment by jenisys** [comment]
Yes, there is a plan to support parallel execution in the future.
BUT: I cannot make any commitments when this functionality will be provided.

Therefore, you are best served to use the suggested URL above (if it helps you).

**Comment by hrcorval** [comment]
Until the native parallel executions are implemented in Behave, I highly recommend checking out the **BehaveX** library. BehaveX is a wrapper that extends the Behave framework with additional features that can significantly improve the performance of your test suites. 
Repo: [https://github.com/hrcorval/behavex](https://github.com/hrcorval/behavex)

Some of the key features include:

 ☑ 𝗣𝗮𝗿𝗮𝗹𝗹𝗲𝗹 test execution for improved performance
 * Easy-to-read HTML test execution 𝗿𝗲𝗽𝗼𝗿𝘁𝘀 with embedded test metrics such as Automation Rate, Pass Rate, and Steps Executions|Duration|Reusability
 ☑ 𝗧𝗲𝘀𝘁 𝗹𝗼𝗴𝘀 per scenario
 ☑ 𝗠𝘂𝘁𝗲 test scenarios in build servers
 ☑ Improved 𝗱𝗿𝘆 𝗿𝘂𝗻𝘀: See the full list of scenarios in an HTML report
 ☑ 𝗥𝗲-𝗲𝘅𝗲𝗰𝘂𝘁𝗲 failing test scenarios

BehaveX is easy to use. Simply install the library and replace the "behave" executable with "behavex" (I recommend taking a look at the documentation as not all Behave arguments are supported yet, but the most important ones). Then, you can run your tests in parallel using the following commands:

▶ 𝚋𝚎𝚑𝚊𝚟𝚎𝚡 -𝚝 @𝚃𝙰𝙶 --𝚙𝚊𝚛𝚊𝚕𝚕𝚎𝚕-𝚙𝚛𝚘𝚌𝚎𝚜𝚜𝚎𝚜 𝟺 --𝚙𝚊𝚛𝚊𝚕𝚕𝚎𝚕-𝚜𝚌𝚑𝚎𝚖𝚎 𝚜𝚌𝚎𝚗𝚊𝚛𝚒𝚘
(𝘛𝘩𝘪𝘴 𝘤𝘰𝘮𝘮𝘢𝘯𝘥 𝘳𝘶𝘯𝘴 𝘵𝘩𝘦 𝘵𝘦𝘴𝘵𝘴 𝘸𝘪𝘵𝘩 𝘵𝘩𝘦 𝘴𝘱𝘦𝘤𝘪𝘧𝘪𝘦𝘥 𝘵𝘢𝘨(𝘴) (@𝘛𝘈𝘎) 𝘪𝘯 𝘱𝘢𝘳𝘢𝘭𝘭𝘦𝘭 𝘶𝘴𝘪𝘯𝘨 4 𝘱𝘳𝘰𝘤𝘦𝘴𝘴𝘦𝘴, 𝘣𝘺 𝘢𝘴𝘴𝘪𝘨𝘯𝘪𝘯𝘨 '𝘴𝘤𝘦𝘯𝘢𝘳𝘪𝘰𝘴' 𝘢𝘤𝘳𝘰𝘴𝘴 𝘱𝘢𝘳𝘢𝘭𝘭𝘦𝘭 𝘱𝘳𝘰𝘤𝘦𝘴𝘴𝘦𝘴)

▶ 𝚋𝚎𝚑𝚊𝚟𝚎𝚡 -𝚝 @𝚃𝙰𝙶 --𝚙𝚊𝚛𝚊𝚕𝚕𝚎𝚕-𝚙𝚛𝚘𝚌𝚎𝚜𝚜𝚎𝚜 𝟹 --𝚙𝚊𝚛𝚊𝚕𝚕𝚎𝚕-𝚜𝚌𝚑𝚎𝚖𝚎 𝚏𝚎𝚊𝚝𝚞𝚛𝚎
(𝘛𝘩𝘪𝘴 𝘤𝘰𝘮𝘮𝘢𝘯𝘥 𝘳𝘶𝘯𝘴 𝘵𝘩𝘦 𝘵𝘦𝘴𝘵𝘴 𝘸𝘪𝘵𝘩 𝘵𝘩𝘦 𝘴𝘱𝘦𝘤𝘪𝘧𝘪𝘦𝘥 𝘵𝘢𝘨(𝘴) (@𝘛𝘈𝘎) 𝘪𝘯 𝘱𝘢𝘳𝘢𝘭𝘭𝘦𝘭 𝘶𝘴𝘪𝘯𝘨 3 𝘱𝘳𝘰𝘤𝘦𝘴𝘴𝘦𝘴, 𝘣𝘺 𝘢𝘴𝘴𝘪𝘨𝘯𝘪𝘯𝘨 '𝘧𝘦𝘢𝘵𝘶𝘳𝘦𝘴' 𝘢𝘤𝘳𝘰𝘴𝘴 𝘱𝘢𝘳𝘢𝘭𝘭𝘦𝘭 𝘱𝘳𝘰𝘤𝘦𝘴𝘴𝘦𝘴)

The --𝗽𝗮𝗿𝗮𝗹𝗹𝗲𝗹-𝗽𝗿𝗼𝗰𝗲𝘀𝘀𝗲𝘀 option specifies the number of processes to use for parallel execution. The --𝗽𝗮𝗿𝗮𝗹𝗹𝗲𝗹-𝘀𝗰𝗵𝗲𝗺𝗲 option specifies the scheme to use for parallel execution. The supported schemes are 'scenario' and 'feature'.

**Comment by carpusherw** [comment]
Is [-j optoin](https://behave.readthedocs.io/en/latest/behave/#cmdoption-j) in 1.2.7 dev releases the official support to parallel execution?

**Reply by jenisys** [reply]
There is currently no support for parallel execution in `behave`.

* Test-runners are an extension-point in `behave`.
* A test-runner that supports parallel execution can use this supported option `—jobs`.

**Reply by carpusherw** [reply]
If I understand correctly, there's currently no `test-runner that supports parallel execution` people can use directly. Instead, people should implement a customized test runner and use read the -j option. Right?
If so, is there an example of how to implement one? I suppose it starts from extending the `run()` of `Runner` class and ends with using `--runner`? I did try to look for documentation and examples but failed to find one. Sorry if I missed something in the docs.
